File: swift/Sources/RemitMd/HttpSigner.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Signer backed by a local HTTP signing server.
///
/// Delegates EIP-712 signing to an HTTP server on localhost (typically
/// `http://127.0.0.1:7402`). The signer server holds the encrypted key;
/// this adapter only needs a bearer token and URL.
///
/// - Bearer token is stored privately, never exposed in description/debug output.
/// - Address is fetched and cached at construction time (GET /address).
/// - `sign(digest:)` POSTs to /sign/digest and returns the hex signature.
/// - All errors are explicit -- no silent fallbacks.
///
/// ```swift
/// let signer = try HttpSigner(url: "http://127.0.0.1:7402", token: "rmit_sk_...")
/// let wallet = RemitWallet(signer: signer, chain: .base)
/// ```
public final class HttpSigner: Signer, @unchecked Sendable {
    public let address: String

    private let url: String
    private let token: String
    private let session: URLSession

    // MARK: - JSON response types

    private struct AddressResponse: Decodable {
        let address: String
    }

    private struct SignatureResponse: Decodable {
        let signature: String?
    }

    private struct ErrorResponse: Decodable {
        let error: String?
        let reason: String?
    }

    // MARK: - Init

    /// Create an HttpSigner, fetching and caching the wallet address from the signer server.
    ///
    /// This constructor is synchronous (required by the `Signer` protocol) and uses
    /// `DispatchSemaphore` to block on the address fetch. This is acceptable for
    /// localhost calls (<1ms latency).
    ///
    /// - Parameters:
    ///   - url: Signer server URL (e.g., "http://127.0.0.1:7402").
    ///   - token: Bearer token for authentication.
    ///   - session: URLSession to use (default: `.shared`). Override in tests.
    /// - Throws: `RemitError` on network errors, auth failures, or missing address.
    public init(url: String, token: String, session: URLSession = .shared) throws {
        self.url = url.hasSuffix("/") ? String(url.dropLast()) : url
        self.token = token
        self.session = session

        // Fetch address synchronously using DispatchSemaphore.
        var request = URLRequest(url: URL(string: self.url + "/address")!)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 10

        let (data, response) = try HttpSigner.syncRequest(request, session: session)
        let http = response as! HTTPURLResponse

        if http.statusCode == 401 {
            throw RemitError(RemitError.unauthorized,
                "HttpSigner: unauthorized -- check your REMIT_SIGNER_TOKEN")
        }
        if http.statusCode == 403 {
            let reason = HttpSigner.parseErrorReason(data)
            throw RemitError(RemitError.unauthorized,
                "HttpSigner: policy denied -- \(reason)")
        }
        guard http.statusCode == 200 else {
            let reason = HttpSigner.parseErrorReason(data)
            throw RemitError(RemitError.serverError,
                "HttpSigner: GET /address failed (\(http.statusCode)): \(reason)")
        }

        guard let addrResp = try? JSONDecoder().decode(AddressResponse.self, from: data),
              !addrResp.address.isEmpty else {
            throw RemitError(RemitError.serverError,
                "HttpSigner: GET /address returned no address")
        }

        self.address = addrResp.address
    }

    // MARK: - Signer protocol

    /// Sign a 32-byte digest by sending it to POST /sign/digest on the signer server.
    ///
    /// - Parameter digest: 32-byte EIP-712 digest.
    /// - Returns: Hex-encoded 65-byte ECDSA signature (r+s+v).
    /// - Throws: `RemitError` on network errors, auth failures, or invalid response.
    public func sign(digest: Data) throws -> String {
        let digestHex = "0x" + digest.hexString

        let body = try JSONEncoder().encode(["digest": digestHex])

        var request = URLRequest(url: URL(string: url + "/sign/digest")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = body
        request.timeoutInterval = 10

        let (data, response) = try HttpSigner.syncRequest(request, session: session)
        let http = response as! HTTPURLResponse

        if http.statusCode == 401 {
            throw RemitError(RemitError.unauthorized,
                "HttpSigner: unauthorized -- check your REMIT_SIGNER_TOKEN")
        }
        if http.statusCode == 403 {
            let reason = HttpSigner.parseErrorReason(data)
            throw RemitError(RemitError.unauthorized,
                "HttpSigner: policy denied -- \(reason)")
        }
        guard http.statusCode == 200 else {
            let reason = HttpSigner.parseErrorReason(data)
            throw RemitError(RemitError.serverError,
                "HttpSigner: POST /sign/digest failed (\(http.statusCode)): \(reason)")
        }

        guard let sigResp = try? JSONDecoder().decode(SignatureResponse.self, from: data),
              let signature = sigResp.signature, !signature.isEmpty else {
            throw RemitError(RemitError.serverError,
                "HttpSigner: server returned no signature")
        }

        return signature
    }

    // MARK: - CustomStringConvertible (no token leakage)

    /// Safe description that never includes the bearer token.
    public var description: String {
        "HttpSigner(address: \(address))"
    }

    // MARK: - Synchronous HTTP helper

    /// Execute a URLRequest synchronously using DispatchSemaphore.
    /// This blocks the calling thread until the response arrives.
    private static func syncRequest(
        _ request: URLRequest,
        session: URLSession
    ) throws -> (Data, URLResponse) {
        let semaphore = DispatchSemaphore(value: 0)
        var resultData: Data?
        var resultResponse: URLResponse?
        var resultError: Error?

        session.dataTask(with: request) { data, response, error in
            resultData = data
            resultResponse = response
            resultError = error
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let error = resultError {
            throw RemitError(RemitError.networkError,
                "HttpSigner: cannot reach signer server at \(request.url?.host ?? "unknown"): \(error.localizedDescription)")
        }
        guard let data = resultData, let response = resultResponse else {
            throw RemitError(RemitError.networkError,
                "HttpSigner: no response from signer server")
        }
        return (data, response)
    }

    // MARK: - Error parsing helper

    /// Extract a human-readable reason from an error response body.
    private static func parseErrorReason(_ data: Data) -> String {
        guard let errResp = try? JSONDecoder().decode(ErrorResponse.self, from: data) else {
            return String(data: data, encoding: .utf8) ?? "unknown"
        }
        return errResp.reason ?? errResp.error ?? "unknown"
    }
}

File: elixir/lib/remit_md/http_signer.ex
defmodule RemitMd.HttpSigner do
  @moduledoc """
  Signer backed by a local HTTP signing server.

  Delegates EIP-712 signing to an HTTP server on localhost (typically
  `http://127.0.0.1:7402`). The signer server holds the encrypted key;
  this adapter only needs a bearer token and URL.

  - Bearer token is stored privately, never exposed via `inspect/1` or `to_string/1`.
  - Address is cached at construction time (`GET /address`).
  - `sign/2` POSTs the raw 32-byte digest to `POST /sign/digest`.
  - All errors are explicit - no silent fallbacks.

  ## Usage

      signer = RemitMd.HttpSigner.new("http://127.0.0.1:7402", "rmit_sk_...")
      wallet = RemitMd.Wallet.new(signer: signer, chain: "base")
  """

  @behaviour RemitMd.Signer

  alias RemitMd.Error

  @enforce_keys [:address, :__url__, :__token__]
  defstruct [:address, :__url__, :__token__]

  @doc """
  Create an HttpSigner by fetching and caching the wallet address from
  `GET /address` on the signer server.

  Raises `RemitMd.Error` on network errors, auth failures, or missing address.
  """
  def new(url, token) when is_binary(url) and is_binary(token) do
    url = String.trim_trailing(url, "/")

    # Ensure :inets and :ssl are started (Application.start handles this,
    # but tests or scripts may not boot the OTP app).
    ensure_http_deps()

    address = fetch_address(url, token)

    %__MODULE__{address: address, __url__: url, __token__: token}
  end

  @impl true
  def sign(%__MODULE__{__url__: url, __token__: token}, digest)
      when is_binary(digest) and byte_size(digest) == 32 do
    ensure_http_deps()

    digest_hex = "0x" <> Base.encode16(digest, case: :lower)
    body = Jason.encode!(%{digest: digest_hex})

    request_url = String.to_charlist(url <> "/sign/digest")
    headers = auth_headers(token)

    result =
      :httpc.request(
        :post,
        {request_url, headers, ~c"application/json", String.to_charlist(body)},
        http_opts(),
        []
      )

    case result do
      {:ok, {{_, 200, _}, _, resp_body}} ->
        parse_signature_response(to_string(resp_body))

      {:ok, {{_, 401, _}, _, _}} ->
        raise Error.new(
          Error.unauthorized(),
          "HttpSigner: unauthorized -- check your REMIT_SIGNER_TOKEN"
        )

      {:ok, {{_, 403, _}, _, resp_body}} ->
        reason = extract_reason(to_string(resp_body))

        raise Error.new(
          Error.forbidden(),
          "HttpSigner: policy denied -- #{reason}"
        )

      {:ok, {{_, status, _}, _, resp_body}} ->
        reason = extract_reason(to_string(resp_body))

        raise Error.new(
          Error.server_error(),
          "HttpSigner: POST /sign/digest failed (#{status}): #{reason}"
        )

      {:error, reason} ->
        raise Error.new(
          Error.network_error(),
          "HttpSigner: cannot reach signer server: #{inspect(reason)}"
        )
    end
  end

  @impl true
  def address(%__MODULE__{address: addr}), do: addr

  # ─── Protocol implementations (never leak token) ────────────────────────────

  defimpl Inspect do
    def inspect(%RemitMd.HttpSigner{address: addr}, _opts) do
      "#RemitMd.HttpSigner<address=#{addr}>"
    end
  end

  defimpl String.Chars do
    def to_string(%RemitMd.HttpSigner{address: addr}) do
      "#RemitMd.HttpSigner<address=#{addr}>"
    end
  end

  # ─── Private ────────────────────────────────────────────────────────────────

  defp fetch_address(url, token) do
    request_url = String.to_charlist(url <> "/address")
    headers = auth_headers(token)

    result = :httpc.request(:get, {request_url, headers}, http_opts(), [])

    case result do
      {:ok, {{_, 200, _}, _, resp_body}} ->
        parse_address_response(to_string(resp_body))

      {:ok, {{_, 401, _}, _, _}} ->
        raise Error.new(
          Error.unauthorized(),
          "HttpSigner: unauthorized -- check your REMIT_SIGNER_TOKEN"
        )

      {:ok, {{_, 403, _}, _, resp_body}} ->
        reason = extract_reason(to_string(resp_body))

        raise Error.new(
          Error.forbidden(),
          "HttpSigner: policy denied -- #{reason}"
        )

      {:ok, {{_, status, _}, _, resp_body}} ->
        reason = extract_reason(to_string(resp_body))

        raise Error.new(
          Error.server_error(),
          "HttpSigner: GET /address failed (#{status}): #{reason}"
        )

      {:error, reason} ->
        raise Error.new(
          Error.network_error(),
          "HttpSigner: cannot reach signer server at #{url}: #{inspect(reason)}"
        )
    end
  end

  defp parse_address_response(body) do
    case Jason.decode(body) do
      {:ok, %{"address" => address}} when is_binary(address) and address != "" ->
        address

      {:ok, _} ->
        raise Error.new(
          Error.server_error(),
          "HttpSigner: GET /address returned no address"
        )

      {:error, _} ->
        raise Error.new(
          Error.server_error(),
          "HttpSigner: GET /address returned invalid JSON"
        )
    end
  end

  defp parse_signature_response(body) do
    case Jason.decode(body) do
      {:ok, %{"signature" => sig}} when is_binary(sig) and sig != "" ->
        sig

      {:ok, _} ->
        raise Error.new(
          Error.server_error(),
          "HttpSigner: server returned no signature"
        )

      {:error, _} ->
        raise Error.new(
          Error.server_error(),
          "HttpSigner: POST /sign/digest returned invalid JSON"
        )
    end
  end

  defp extract_reason(body) do
    case Jason.decode(body) do
      {:ok, %{"reason" => reason}} when is_binary(reason) and reason != "" ->
        reason

      {:ok, %{"error" => error}} when is_binary(error) and error != "" ->
        error

      _ ->
        if body != "" and String.length(body) < 200, do: body, else: "unknown"
    end
  end

  defp auth_headers(token) do
    [{~c"authorization", String.to_charlist("Bearer #{token}")}]
  end

  defp http_opts do
    [timeout: 10_000, connect_timeout: 5_000]
  end

  defp ensure_http_deps do
    case :inets.start() do
      :ok -> :ok
      {:error, {:already_started, :inets}} -> :ok
    end

    case :ssl.start() do
      :ok -> :ok
      {:error, {:already_started, :ssl}} -> :ok
    end
  end
end
